stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -0,0 +1,132 @@
+ *  whether to continue the recursion or to perform the cholesky.
+ * @return Square root of matrix on the GPU.
+ * @throw std::domain_error if m is not
+ *  positive definite (if m has more than 0 elements)
+ */
+inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int block = 100,
+                                     const int divider = 2,
+                                     const int min_block = 100) {
+  auto offset = 0;
+  // NOTE: The code in this section follows the naming conventions
+  // in the report linked in the docs.
+  matrix_gpu A_11(block, block);
+  matrix_gpu L_11(block, block);
+  // Repeats the blocked cholesky decomposition until the size of the remaining
+  // submatrix is smaller or equal to the block size
+  while ((offset + block) < (A.rows())) {
+    auto block_subset = A.rows() - offset - block;
+    matrix_gpu A_21(block_subset, block);
+    matrix_gpu A_22(block_subset, block_subset);
+    // Copies the A_11 submatrix from the input
+    A_11.sub_block(A, offset, offset, 0, 0, block, block);
+    // Calls the blocked cholesky for the submatrix A_11
+    // or calls the kernel  directly if the size of the block is small enough
+    if (block <= min_block || divider <= 1) {
+      L_11.zeros();
+      try {
+        opencl_kernels::cholesky_decompose(
+            cl::NDRange(A_11.rows()), cl::NDRange(A_11.rows()), A_11.buffer(),
+            L_11.buffer(), A_11.rows());
+      } catch (const cl::Error& e) {
+        check_opencl_error("cholesky_decompose", e);
+      }
+    } else {
+      L_11 = cholesky_decompose(A_11, block / divider, divider, min_block);
+    }
+    // Copies the cholesky factor of A_11 back to the input matrix
+    A.sub_block(L_11, 0, 0, offset, offset, block, block);
+    // Copies the A_21 submatrix from the input matrix,
+    auto block_offset = offset + block;
+    A_21.sub_block(A, block_offset, offset, 0, 0, block_subset, block);
+    // computes A_21*((L_11^-1)^T)
+    // and copies the resulting submatrix to the input matrix
+    matrix_gpu A_11_inverse = lower_triangular_inverse(L_11);
+    A_11_inverse = transpose(A_11_inverse);
+    // TODO(Steve): Replace with mult operator when that PR goes through
+    matrix_gpu L_21 = multiply(A_21, A_11_inverse);
+    A.sub_block(L_21, 0, 0, block_offset, offset, block_subset, block);
+    // Copies the A_22 submatrix from the input matrix,
+    A_22.sub_block(A, block_offset, block_offset, 0, 0, block_subset,
+                   block_subset);
+    // computes A_22 - L_21*(L_21^T)
+    // and copies the resulting submatrix back to the input matrix
+    matrix_gpu temp = multiply_transpose(L_21);
+    // TODO(Steve): Replace with subtraction operator when that PR goes through
+    matrix_gpu L_22 = subtract(A_22, temp);
+    A.sub_block(L_22, 0, 0, block_offset, block_offset, block_subset,
+                block_subset);
+    offset += block;
+  }
+  // Computes the Cholesky factor for the remaining part of the matrix
```

Comment:
This looks familiar... Any way we can unify this with previous code in this method that looks similar?

Reply:
If I write a `cholesky_decompose_recursion` method should it go in the internal namespace? Since idt we want users to touch a function like that.

Reply:
yep, that sounds good.